### PR TITLE
Added support quick-jumping to web-based documentation 

### PR DESCRIPTION
**Pull request by obicons**
When working on Terraform, I commonly need to jump to documentation. I find myself jumping back and forth between Terraform's documentation and Emacs frequently. 

This code adds support for hitting "M-." while the point is positioned at a resource to automatically navigate to Terraform's web-based documentation.

### CODE REVIEW

File: terraform-mode.el [hunk 1]
```diff
@@ -85,6 +86,37 @@
...
+  (car (split-string resource-name "_")))
+
+(defun resource (resource-name)
+  "Returns the resource associated with a resource-name"
+  (apply #'concat
+         (interpolate (cdr (split-string resource-name "_")) "_")))
+
+(defun terraform-resource-url (resource)
+  "Returns the url containing the documentation for resource"
+  (format "https://www.terraform.io/docs/providers/%s/r/%s.html"
```
Comment (rnikoopour) [inline terraform-mode.el:107]: This URL format does a 301 redirect to a new URL format.  Ideally this should use the new URL format.

File: terraform-mode.el [hunk 2]
```diff
@@ -85,6 +86,37 @@
...
+                   (interpolate (cdr list) item)))))
+
+(defun provider (resource-name)
+  "Returns the provider associated with a resource-name"
+  (car (split-string resource-name "_")))
+
+(defun resource (resource-name)
+  "Returns the resource associated with a resource-name"
+  (apply #'concat
+         (interpolate (cdr (split-string resource-name "_")) "_")))
```
Comment (rnikoopour) [inline terraform-mode.el:103]: There's no need to write your own `interpolate`

```
(defun terraform--get-resource-type-without-provider (resource-type)
  "Returns a resource type without its provider prefix"
  (string-join (cdr (split-string resource-type "_"))
	       "_"))
```

File: terraform-mode.el [hunk 3]
```diff
@@ -85,6 +86,37 @@
...
     (remove-hook 'before-save-hook #'terraform-format-buffer t)))
 
+(defun interpolate (list item)
+  "Places item between elements in list"
+  (cond ((null list) nil)
+        ((= (length list) 1) list)
+        (t (append (list (car list) item)
+                   (interpolate (cdr list) item)))))
+
+(defun provider (resource-name)
```
Comment (rnikoopour) [inline terraform-mode.el:96]: Please rename to `terraform--get-provider-from-resource-type (resource-type)`

File: terraform-mode.el [hunk 4]
```diff
@@ -85,6 +86,37 @@
...
+(defun terraform-resource-url (resource)
+  "Returns the url containing the documentation for resource"
+  (format "https://www.terraform.io/docs/providers/%s/r/%s.html"
+          (provider resource)
+          (resource resource)))
+
+(defun terraform-open-doc ()
+  "Opens a browser at the URL documenting the resource-at-point"
+  (interactive)
+  (browse-url (terraform-resource-url (thing-at-point 'symbol))))
```
Comment (rnikoopour) [inline terraform-mode.el:114]: This looks like it requires the user to have the point hovering over the resource-type for the block.  Would there be a way to have this auto-navigate to the resource type given it's called from within a resource block?

**Review by rnikoopour — CHANGES_REQUESTED** [review]

**Comment by dod38fr** [comment]
For team work in a team where not all people use emacs, it would also be useful to add a comment containing a link to documentation above the resource. I.e. something like:

```terraform
# https://github.com/cloudposse/terraform-aws-tfstate-backend#usage
module "terraform_state_backend" {
  source = "cloudposse/tfstate-backend/aws"
```

where the comment was added by terraform-mode.

File: terraform-mode.el (hunk 1 above)
Comment (dod38fr) [reply terraform-mode.el:107]: Latest URLs format includes the namespace of the provider (e.g. `hashiscorp`). For instance:
```
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudfront_cache_policy
```

This information can be retrieved in `terraform` block, which may be in another `tf` file in the same directory. Here and example:
```terraform
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 4.8"
    }
  }
}
```

I'm an elisp noob, so I don't know how to get this information ...

**Comment by dod38fr** [comment]
@Fuco1 Thanks a bunch. I'll test this next week :-)

**Pull request merged**